File: src/my_py_pkg/my_py_pkg/count_until_client.py
import rclpy
import time
from rclpy.node import Node
from my_robot_interfaces.action import CountUntil
from rclpy.action import ActionClient
from action_msgs.msg import GoalStatus

class CountUntilActionClient(Node):

    def __init__(self):
        super().__init__('count_until_client')
        self._action_client = ActionClient(self, CountUntil, 'count_until')
        self._sent_goals = 0
        self._received_results = 0
        self._goal_handles = {}

    def send_goal(self, target_number, period, cancel=False):
        goal_msg = CountUntil.Goal()
        goal_msg.target_number = target_number
        goal_msg.period = period

        self._action_client.wait_for_server()
        goal_id = self._sent_goals
        self._sent_goals += 1
        self._send_goal_future = self._action_client.send_goal_async(goal_msg, feedback_callback=self.feedback_callback)
        self._send_goal_future.add_done_callback(lambda future: self.goal_response_callback(future, goal_id))

        # Send a cancel request after 2 seconds
        if cancel:
            self._timer = self.create_timer(2.0, lambda: self.send_cancel_request(goal_id))
        
        
    def send_cancel_request(self, goal_id):
        if not self._goal_handles[goal_id] is None and self._goal_handles[goal_id].status == GoalStatus.STATUS_EXECUTING:
            self.get_logger().info(f'Sending cancel request with argument goal_id={goal_id}')
            self._goal_handles[goal_id].cancel_goal_async() # Cancel the goal

        self._timer.cancel() # Cancel the timer
        
    def goal_finished(self):
        self.get_logger().info('Goal finished!')
        self._received_results += 1
        if self._sent_goals == self._received_results:
            self.get_logger().info('All goals are done!')
            rclpy.shutdown()


    def goal_response_callback(self, future, goal_id):
        self._goal_handles[goal_id] = future.result()
        if not self._goal_handles[goal_id].accepted:
            self.get_logger().warn('Goal rejected :(')
            self.goal_finished()
            return

        self.get_logger().info('Goal accepted :)')
        self._get_result_future = self._goal_handles[goal_id].get_result_async()
        self._get_result_future.add_done_callback(self.get_result_callback)

    def get_result_callback(self, future):
        result = future.result().result
        status = future.result().status

        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info('Goal succeeded!')
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().warn('Goal was canceled')
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error('Goal was aborted')
        else:
            self.get_logger().warn('Goal failed with status: {0}'.format(status))
        
        self.get_logger().info('Result: {0}'.format(result.reached_number))
        
        self.goal_finished()

    def feedback_callback(self, feedback_msg):
        feedback = feedback_msg.feedback
        self.get_logger().info('Received feedback: {0}'.format(feedback.current_number))

def main(args=None):
    rclpy.init(args=args)
    action_client = CountUntilActionClient()
    action_client.send_goal(6, 1.0, True)
    action_client.send_goal(10, 0.5)
    rclpy.spin(action_client)

if __name__ == '__main__':
    main()

File: src/actions_cpp/src/count_until_client.cpp
#include <chrono>
#include <map>
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include "my_robot_interfaces/action/count_until.hpp"

using namespace std::chrono_literals;

class CountUntilClient : public rclcpp::Node
{
public:
    using CountUntil = my_robot_interfaces::action::CountUntil;
    using GoalHandleCountUntil = rclcpp_action::ClientGoalHandle<CountUntil>;

    explicit CountUntilClient(const rclcpp::NodeOptions & options = rclcpp::NodeOptions())
    : Node("count_until_client", options)
    {
        this->client_ptr_ = rclcpp_action::create_client<CountUntil>(this, "count_until");
    }

    void send_goal(int target_number, float period, bool cancel = false)
    {
        using namespace std::placeholders;

        // Wait for action server to be up and active
        if (!this->client_ptr_->wait_for_action_server()) {
            RCLCPP_ERROR(this->get_logger(), "Action server not available after waiting");
            rclcpp::shutdown();
            return;
        }

        // Create a goal message
        auto goal_msg = CountUntil::Goal();
        goal_msg.target_number = target_number;
        goal_msg.period = period;

        // Send the goal
        RCLCPP_INFO(this->get_logger(), "Sending goal with target_number %d and period %.2f", target_number, period);

        int goal_id = sent_goals_num_;
        auto send_goal_options = rclcpp_action::Client<CountUntil>::SendGoalOptions();
        send_goal_options.goal_response_callback =
            std::bind(&CountUntilClient::goal_response_callback, this, _1, goal_id);
        send_goal_options.feedback_callback =
            std::bind(&CountUntilClient::feedback_callback, this, _1, _2);
        send_goal_options.result_callback =
            std::bind(&CountUntilClient::result_callback, this, _1);

        this->client_ptr_->async_send_goal(goal_msg, send_goal_options);
        sent_goals_num_++;
        if(cancel)
        {
            timer_ = this->create_wall_timer(2s, [this, target_number, period, goal_id]() 
            {
                if(goal_handles_.find(goal_id) != goal_handles_.end() && goal_handles_[goal_id] && goal_handles_[goal_id]->get_status() == action_msgs::msg::GoalStatus::STATUS_EXECUTING)
                {
                    RCLCPP_INFO(this->get_logger(), "Cancelling goal with target_number %d and period %.2f", target_number, period);
                    this->client_ptr_->async_cancel_goal(goal_handles_[goal_id]);
                }
                timer_->cancel();
            });
        }
    }

private:
    rclcpp_action::Client<CountUntil>::SharedPtr client_ptr_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::unordered_map<int, GoalHandleCountUntil::SharedPtr> goal_handles_;

    int received_goals_num_ = 0;
    int sent_goals_num_ = 0;

    void goal_finished_callback()
    {
        RCLCPP_INFO(this->get_logger(), "Goal finished");
        received_goals_num_++;

        // Shutdown the node after receiving the result
        if(sent_goals_num_ == received_goals_num_)
            rclcpp::shutdown();
    }   

    // Callback to receive the goal response, i.e. goal was accepted or rejected
    void goal_response_callback(const GoalHandleCountUntil::SharedPtr& goal_handle, const int goal_id)
    {
        if (!goal_handle) {
            RCLCPP_ERROR(this->get_logger(), "Goal %d was rejected by server", goal_id);
            goal_finished_callback();
        } else {
            goal_handles_[goal_id] = goal_handle;
            RCLCPP_INFO(this->get_logger(), "Goal %d accepted by server, waiting for result", goal_id);
        }
    }

    // Callback to receive feedback
    void feedback_callback(
        GoalHandleCountUntil::SharedPtr,
        const std::shared_ptr<const CountUntil::Feedback> feedback)
    {
        RCLCPP_INFO(this->get_logger(), "Got feedback: %ld", feedback->current_number);
    }

    // Callback to receive the result once the goal is finished
    void result_callback(const GoalHandleCountUntil::WrappedResult & result)
    {
        switch (result.code) 
        {
            case rclcpp_action::ResultCode::SUCCEEDED:
                RCLCPP_INFO(this->get_logger(), "Result received: %ld", result.result->reached_number);
                break;
            case rclcpp_action::ResultCode::ABORTED:
                RCLCPP_ERROR(this->get_logger(), "Goal was aborted");
                break;
            case rclcpp_action::ResultCode::CANCELED:
                RCLCPP_ERROR(this->get_logger(), "Goal was canceled");
                break;
            default:
                RCLCPP_ERROR(this->get_logger(), "Unknown result code");
                break;
        }
        goal_finished_callback();
    }
};

int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CountUntilClient>();
    node->send_goal(6, 1.0, false);
    node->send_goal(-2, 0.6);
    node->send_goal(20, 0.4);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}